File: bshop/src/main/java/com/example/bshop/service/ProduitService.java
package com.example.bshop.service;

import com.example.bshop.model.Produit;
import com.example.bshop.repository.ProduitRepository;
import org.springframework.stereotype.Service;
import java.util.List;
import java.util.Optional;

@Service
public class ProduitService {

    private final ProduitRepository produitRepository;

    public ProduitService(ProduitRepository produitRepository) {
        this.produitRepository = produitRepository;
    }

    public Produit create(Produit produit) {
        // Vérifier si un produit avec le même nom existe déjà
        Optional<Produit> existing = produitRepository.findByNom(produit.getNom());
        if (existing.isPresent()) {
            throw new RuntimeException("Un produit avec le nom '" + produit.getNom() + "' existe déjà");
        }
        return produitRepository.save(produit);
    }

    public List<Produit> readAll() {
        return produitRepository.findAll();
    }

    public Optional<Produit> read(Long id) {
        return produitRepository.findById(id);
    }

    public Produit update(Long id, Produit produitModifie) {
        // Vérifier si un autre produit avec le même nom existe
        Optional<Produit> existing = produitRepository.findByNom(produitModifie.getNom());
        if (existing.isPresent() && !existing.get().getId().equals(id)) {
            throw new RuntimeException("Un produit avec le nom '" + produitModifie.getNom() + "' existe déjà");
        }

        Produit produit = produitRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Produit non trouvé : " + id));

        produit.setNom(produitModifie.getNom());
        produit.setDescription(produitModifie.getDescription());
        produit.setPrix(produitModifie.getPrix());
        produit.setPrixOriginal(produitModifie.getPrixOriginal());
        produit.setStock(produitModifie.getStock());
        produit.setImageUrl(produitModifie.getImageUrl());
        produit.setCategorie(produitModifie.getCategorie());
        produit.setSousCategorie(produitModifie.getSousCategorie());
        produit.setGenre(produitModifie.getGenre());
        produit.setRating(produitModifie.getRating());
        produit.setIsPromo(produitModifie.getIsPromo());
        produit.setIsNew(produitModifie.getIsNew());
        produit.setTailles(produitModifie.getTailles());
        produit.setCouleurs(produitModifie.getCouleurs());

        return produitRepository.save(produit);
    }

    public void delete(Long id) {
        produitRepository.deleteById(id);
    }

    // Supprimer plusieurs produits
    public void deleteAll(List<Long> ids) {
        produitRepository.deleteAllById(ids);
    }

    public List<Produit> findByCategorie(String categorie) {
        return produitRepository.findByCategorie(categorie);
    }

    public List<Produit> findBySousCategorie(String sousCategorie) {
        return produitRepository.findBySousCategorie(sousCategorie);
    }

    public List<Produit> findByGenre(String genre) {
        return produitRepository.findByGenre(genre);
    }

    public List<Produit> search(String keyword) {
        return produitRepository.searchByKeyword(keyword);
    }
}

File: fshop/src/service/produit.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Produit } from '../model/produit';

@Injectable({
  providedIn: 'root'
})
export class ProduitService {
  private apiUrl = 'http://localhost:9091/api/produits';

  constructor(private http: HttpClient) {}

  create(produit: Produit): Observable<Produit> {
    return this.http.post<Produit>(`${this.apiUrl}/create`, produit);
  }

  readAll(): Observable<Produit[]> {
    return this.http.get<Produit[]>(`${this.apiUrl}/readall`);
  }

  read(id: number): Observable<Produit> {
    return this.http.get<Produit>(`${this.apiUrl}/read/${id}`);
  }

  update(id: number, produit: Produit): Observable<Produit> {
    return this.http.put<Produit>(`${this.apiUrl}/update/${id}`, produit);
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/delete/${id}`);
  }

  findByCategorie(categorie: string): Observable<Produit[]> {
    return this.http.get<Produit[]>(`${this.apiUrl}/categorie/${categorie}`);
  }

  findBySousCategorie(sousCategorie: string): Observable<Produit[]> {
    return this.http.get<Produit[]>(`${this.apiUrl}/sous-categorie/${sousCategorie}`);
  }

  findByGenre(genre: string): Observable<Produit[]> {
    return this.http.get<Produit[]>(`${this.apiUrl}/genre/${genre}`);
  }

  search(keyword: string): Observable<Produit[]> {
    const params = new HttpParams().set('keyword', keyword);
    return this.http.get<Produit[]>(`${this.apiUrl}/search`, { params });
  }

  importExcel(file: File): Observable<Produit[]> {
  const formData = new FormData();
  formData.append('file', file);
  return this.http.post<Produit[]>(`${this.apiUrl}/import-excel`, formData);
}
}